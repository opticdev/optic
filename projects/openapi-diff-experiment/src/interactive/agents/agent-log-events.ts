import { OpenAPIV3 } from '@useoptic/openapi-utilities';

export type AgentLogEvent =
  | ReadingSpec
  | PatchingSpec
  | UnexpectedError
  | DiffingTraffic;

export enum AgentLogEvents {
  reading,
  patching,
  error,
  diffingTraffic,
}

type ReadingSpec = { event: AgentLogEvents.reading; location: string };
type PatchingSpec = { event: AgentLogEvents.patching; patches: string[] };
type UnexpectedError = {
  event: AgentLogEvents.error;
  error: string;
  classification: string;
};
type DiffingTraffic = {
  // matched?: { pathPattern: string; method: OpenAPIV3.HttpMethods };
  path: string;
  hasDiffs: boolean;
  statusCode: string;
  method: string;
  event: AgentLogEvents.diffingTraffic;
};
