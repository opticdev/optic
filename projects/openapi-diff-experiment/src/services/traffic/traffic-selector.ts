import { ApiTraffic } from './types';
import { ConceptualLocation } from '@useoptic/openapi-utilities';
import invariant from 'ts-invariant';

export function trafficSelector(traffic: ApiTraffic) {
  return {
    bodyJsonForLocation: (location: ConceptualLocation) => {
      if ('inResponse' in location && 'body' in location.inResponse) {
        const contentType = location.inResponse.body.contentType;

        invariant(
          contentType === traffic.response.body.contentType,
          'content type must match when selecting examples from traffic'
        );

        invariant(
          typeof traffic.response.body.jsonBodyString !== 'undefined',
          'body json must be present if selecting example from traffic'
        );

        return JSON.parse(traffic.response.body.jsonBodyString!);
      }

      if ('inRequest' in location && 'body' in location.inRequest) {
        const { contentType } = location.inRequest.body;
        invariant(
          contentType === traffic.requestBody?.contentType,
          'content type must match when selecting examples from traffic'
        );
        invariant(
          typeof traffic.requestBody.jsonBodyString !== 'undefined',
          'body json must be present if selecting example from traffic'
        );
        return JSON.parse(traffic.requestBody.jsonBodyString!);
      }

      invariant(false, 'can not select example body from traffic');
    },
  };
}
