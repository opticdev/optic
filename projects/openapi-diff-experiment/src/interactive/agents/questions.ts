/*
  Agents are presented with questions, and must decide how to answer them. Possible answers
  - see a new url | add a new path, ignore the path/method pattern
  - see a shape diff | extend spec, replace, type
  - see a new body   | add it, ignore it
 */

import { ApiTraffic } from '../../services/traffic/types';
import {
  DiffType,
  IDiff,
  ShapeDiffTypes,
  UnmatchedPath,
} from '../../services/diff/types';
import { v4 as uuidv4 } from 'uuid';
import { IPatchGroup } from '../../services/patch/incremental-json-patch/json-patcher';
import { JsonSchemaPatch } from '../../services/diff/differs/json-schema-json-diff/plugins/plugin-types';

export type QuestionsForAgent = AddPathQuestion | PatchesForBodyPropertyDiff;

export type QuestionsForAgentAnswered = QuestionsForAgent & {
  answer: QuestionsForAgent['answer'];
};

interface Questions<A> {
  type: AnswerQuestionTypes;
  answer?: A;
  uuid: string;
}

export enum AnswerQuestionTypes {
  AddPath,
  ReviewPatchesForBodyPropertyDiff,
}

// Path add  Q / A
export interface AddPathQuestion extends Questions<AddPathAnswer> {
  type: AnswerQuestionTypes.AddPath;
  diff: UnmatchedPath;
  example: ApiTraffic;
}
export interface AddPathAnswer {
  pathPattern: string;
}

// Body Property add  Q / A
export interface PatchesForBodyPropertyDiff
  extends Questions<PatchesForBodyPropertyDiffAnswer> {
  type: AnswerQuestionTypes.ReviewPatchesForBodyPropertyDiff;
  diff: ShapeDiffTypes;
  example: ApiTraffic;
}
export interface PatchesForBodyPropertyDiffAnswer {
  patch: JsonSchemaPatch;
}
