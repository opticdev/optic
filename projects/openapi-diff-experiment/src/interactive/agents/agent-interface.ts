import { AnswerQuestionTypes, QuestionsForAgent } from './questions';
import { ApiTraffic } from '../../services/traffic/types';
import { IDiff, IDiffService } from '../../services/diff/types';
import { IPatchOpenAPI } from '../../services/patch/types';
import { OpenAPIDiffingQuestions } from '../../services/read/types';
import { OpenAPIV3 } from 'openapi-types';

export interface AgentContext {}

export enum AgentEventEnum {
  DiffReceived = 'DiffReceived',
  Reset = 'Reset',
  AnswerQuestion = 'AnswerQuestion',
  SkipInteraction = 'SkipInteraction',
  SkipQuestion = 'SkipQuestion',
  AdvancePastError = 'AdvancePastError',
  Start = 'Start',
  Finish = 'Finish',
  DisplayError = 'DisplayError',
  BypassError = 'BypassError',
}

export type ProcessDiffContext = AgentContext & {
  diffs: IDiff[];
  traffic: ApiTraffic;
};
export type WaitingOnInputDiffContext = AgentContext & {
  diffs: IDiff[];
  questions: QuestionsForAgent[];
  traffic: ApiTraffic;
};

export type AgentEvent =
  | { type: AgentEventEnum.DiffReceived; diffs: IDiff[]; traffic: ApiTraffic }
  | { type: AgentEventEnum.AnswerQuestion; id: string; answer: any }
  | { type: AgentEventEnum.Reset }
  | { type: AgentEventEnum.SkipInteraction }
  | { type: AgentEventEnum.SkipQuestion }
  | { type: AgentEventEnum.DisplayError }
  | { type: AgentEventEnum.BypassError };

export type AgentTypestate =
  | {
      value: 'idle';
      context: AgentContext;
    }
  | {
      value: 'process_diffs';
      context: ProcessDiffContext;
    }
  | {
      value: 'waiting_for_input';
      context: WaitingOnInputDiffContext;
    }
  | {
      value: 'error';
      context: AgentContext & { error: string };
    }
  | {
      value: 'finished';
      context: AgentContext & {};
    };

export interface AgentIntent {
  name: string;
  filterDiffs: (
    diffs: IDiff[],
    questions: OpenAPIDiffingQuestions,
    traffic: ApiTraffic,
    spec: OpenAPIV3.Document
  ) => IDiff[];
  handleDiffs: (
    diff: IDiff,
    example: ApiTraffic,
    patch: IPatchOpenAPI,
    diffService: IDiffService,
    askQuestion: (question: QuestionsForAgent) => void
  ) => void;
  // filterDiffs:
  applyAnswerAsPatch: (
    questionAnswer: QuestionsForAgent,
    patch: IPatchOpenAPI
  ) => void;
}
