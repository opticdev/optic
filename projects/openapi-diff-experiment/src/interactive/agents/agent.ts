import { assign, createMachine, interpret } from 'xstate';
import {
  AgentContext,
  AgentEvent,
  AgentEventEnum,
  AgentIntent,
  AgentTypestate,
  ProcessDiffContext,
  WaitingOnInputDiffContext,
} from './agent-interface';
import isEqual from 'lodash.isequal';
import { InteractiveDiffMachineType } from '../machine';
import { IDiff } from '../../services/diff/types';
import { QuestionsForAgent } from './questions';
import { ApiTraffic } from '../../services/traffic/types';
import { DiffEventEnum } from '../machine-interface';

export function createAgentMachine(
  diffMachine: InteractiveDiffMachineType,
  intent: AgentIntent
) {
  const agentMachine = createMachine<AgentContext, AgentEvent, AgentTypestate>({
    id: 'agent',
    initial: 'idle',
    context: {},
    states: {
      idle: {
        entry: assign({}),
        on: {
          [AgentEventEnum.DisplayError]: {
            target: 'error',
          },
          [AgentEventEnum.DiffReceived]: {
            target: 'process_diffs',
            actions: assign((context, event) => {
              return {
                diffs: event.diffs,
                traffic: event.traffic,
                questions: [],
              };
            }),
          },
        },
      },
      error: {
        on: {
          [AgentEventEnum.BypassError]: {
            actions: (ctx, event) =>
              diffMachine.send({ type: DiffEventEnum.Bypass_Error }),
            target: 'idle',
          },
        },
      },
      process_diffs: {
        invoke: {
          // this entire src could be made a service for other kinds of diff intents
          src: async (
            ctx,
            event
          ): Promise<{ questions: QuestionsForAgent[] }> => {
            const { diffs, traffic } = ctx as ProcessDiffContext;

            // for user input
            const questions: QuestionsForAgent[] = [];
            // use this to patch diffs this intent automates
            const { specInterface, diffService } = diffMachine.state.context;

            const filteredDiffs = intent.filterDiffs(
              diffs,
              await specInterface.read.questions(),
              traffic,
              await specInterface.read.flattenedSpecification()
            );

            filteredDiffs.forEach((diff) =>
              intent.handleDiffs(
                diff,
                traffic,
                specInterface.patch,
                diffService,
                (question) => {
                  questions.push(question);
                }
              )
            );

            return { questions };
          },
          onDone: [
            {
              target: 'check_should_flush',
              actions: assign((ctx, event) => {
                return {
                  questions: event.data.questions,
                };
              }),
            },
          ],
          onError: {
            actions: (ctx, event) => {
              console.log(event);
            },
          },
        },
      },
      waiting_for_input: {
        on: {
          [AgentEventEnum.DisplayError]: {
            target: 'error',
          },
          [AgentEventEnum.Reset]: {
            target: 'idle',
          },
          [AgentEventEnum.SkipInteraction]: {
            actions: (ctx, event) => {
              diffMachine.send({
                type: DiffEventEnum.Agent_Skipped_Interaction,
              });
            },
            target: 'idle',
          },
          [AgentEventEnum.SkipQuestion]: {
            actions: assign((ctx) => {
              const newQuestions = (ctx as WaitingOnInputDiffContext).questions;
              newQuestions.shift();
              return { questions: newQuestions };
            }),
            target: 'check_should_flush',
          },
          [AgentEventEnum.AnswerQuestion]: {
            actions: assign((ctx, event) => {
              const context = ctx as WaitingOnInputDiffContext;
              const { specInterface } = diffMachine.state.context;
              const question = context.questions.find(
                (i) => i.uuid === event.id
              );
              if (question) {
                question.answer = event.answer;
              }
              intent.applyAnswerAsPatch(question, specInterface.patch);
              return { questions: context.questions };
            }),
            target: 'check_should_flush',
          },
        },
      },
      check_should_flush: {
        invoke: {
          src: async (ctx) => {
            const context = ctx as WaitingOnInputDiffContext;
            const { specInterface } = diffMachine.state.context;

            const result = { shouldIdle: true, shouldFlush: false };

            if (
              context.questions.length > 0 &&
              context.questions.some((i) => i.answer)
            ) {
              result.shouldFlush = true;
            }

            if (
              context.questions.length > 0 &&
              context.questions.every((i) => !i.answer)
            ) {
              result.shouldFlush = false;
              result.shouldIdle = false;
            }

            if (
              context.questions.length === 0 &&
              specInterface.patch.listPatches().length > 0
            ) {
              result.shouldFlush = true;
            }

            return result;
          },
          onDone: [
            {
              target: 'idle',
              actions: (ctx) => {
                const additionalQuestions = ctx as WaitingOnInputDiffContext;
                const dropCurrentTraffic = !additionalQuestions.questions.some(
                  (i) => !i.answer
                );

                diffMachine.send({
                  type: DiffEventEnum.Agent_Submitted_Patch,
                  dropCurrentTraffic,
                });
              },
              cond: (ctx, event) => event.data.shouldFlush,
            },
            {
              target: 'idle',
              cond: (ctx, event) => event.data.shouldIdle,
            },
            { target: 'waiting_for_input' },
          ],
          onError: {
            actions: [
              (ctx, error) => {
                diffMachine.send({
                  type: DiffEventEnum.Agent_Skipped_Interaction,
                });
              },
            ],
            target: 'idle',
          },
        },
      },
    },
  });

  const service = interpret(agentMachine);

  service.start();

  // response to state changes of diff machine
  let previousAsk: { diffs: IDiff[]; traffic: ApiTraffic } | undefined;

  diffMachine.onTransition((state, event) => {
    // transitioned to waiting_for_input

    if (typeof state.value === 'string' && state.value.endsWith('_error')) {
      return service.send({
        type: AgentEventEnum.DisplayError,
      });
    }

    if (state.value === 'waiting_for_input') {
      if (
        !isEqual(previousAsk, {
          diffs: state.context.diffs,
          traffic: state.context.queue[0],
        })
      ) {
        service.send({
          type: AgentEventEnum.DiffReceived,
          diffs: state.context.diffs,
          traffic: state.context.queue[0],
        });
        previousAsk = {
          diffs: state.context.diffs,
          traffic: state.context.queue[0],
        };
      }
    } else {
      previousAsk = undefined;
      service.send({
        type: AgentEventEnum.Reset,
      });
    }
  });

  return { agent: service, diffMachine: diffMachine };
}
