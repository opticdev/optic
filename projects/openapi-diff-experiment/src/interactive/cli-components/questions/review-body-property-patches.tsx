import React, { useMemo } from 'react';
import { Box, Newline } from 'ink';
// @ts-ignore
import Divider from 'ink-divider';
import { PatchesForBodyPropertyDiff } from '../../agents/questions';
import { ChangeView } from '../ui/change';
import { DiffType } from '../../../services/diff/types';
import chalk from 'chalk';
import { CodeBlock } from '../ui/code/code-block';
import invariant from 'ts-invariant';
import { usePatches } from '../hooks/use-patches';
import { useDiffAgentActions } from '../context/diff-agent-context';

const options = [
  {
    char: 'a',
    action: 'add path and operation',
  },
  {
    char: 's',
    action: 'skip this example',
  },
];

export function ReviewBodyPropertyPatches(props: {
  question: PatchesForBodyPropertyDiff;
}) {
  const [diffText, diffVisualization] = useMemo(
    () => diffToPreview(props.question),
    [props.question]
  );

  const { loading, patches } = usePatches(props.question.diff);
  const actions = useDiffAgentActions();

  return (
    <Box flexDirection={'column'}>
      <Newline />
      <ChangeView
        onAnswer={(patch) => {
          actions.answer(props.question.uuid, { patch });
        }}
        patches={{ loading, previews: patches }}
        observation={diffText}
      >
        {diffVisualization}
      </ChangeView>
    </Box>
  );
}

// helpers
function diffToPreview(
  question: PatchesForBodyPropertyDiff
): [string, React.ReactElement] {
  const example = (() => {
    if ('inRequest' in question.diff.location) {
      // @todo fix me -- need request version
      throw new Error('request needs to be implimented');
    } else if ('inResponse' in question.diff.location) {
      return JSON.parse(question.example.response.body.jsonBodyString);
    } else {
      invariant(false, 'example for diff can not be found');
    }
  })();

  switch (question.diff.type) {
    case DiffType.BodyAdditionalProperty:
      return [
        `Additional property ${chalk.cyan.bold(question.diff.key)}`,
        <CodeBlock
          render={{
            json: example,
            highlight: {
              highlight: 'green',
              trail: question.diff.instancePath,
            },
          }}
        />,
      ];
    case DiffType.BodyMissingRequiredProperty:
      return [
        `Missing required property ${chalk.cyan.bold(question.diff.key)}`,
        <CodeBlock
          render={{
            json: example,
            highlight: {
              highlight: 'red',
              wasMissing: true,
              trail: question.diff.instancePath,
            },
          }}
        />,
      ];
    case DiffType.BodyUnmatchedType:
      return [
        `Unmatched type for property ${chalk.cyan.bold(question.diff.key)}`,
        <CodeBlock
          render={{
            json: example,
            highlight: {
              highlight: 'yellow',
              trail: question.diff.instancePath,
            },
          }}
        />,
      ];
  }
}
