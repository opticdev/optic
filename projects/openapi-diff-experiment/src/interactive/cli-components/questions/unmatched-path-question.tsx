import React, { useMemo, useState } from 'react';
import { Box, Newline, Text, useInput } from 'ink';
import TextInput from 'ink-text-input';
// @ts-ignore
import Divider from 'ink-divider';
import { RenderPathMethod } from '../ui/render-path-method';
import {
  urlPathDiff,
  urlPathDiffFromSpec,
} from '../../../services/diff/differs/url-path-diff';
import quote from 'ajv/dist/runtime/quote';
import { AnswerBar } from '../ui/answer-bar';
import { AddPathAnswer, AddPathQuestion } from '../../agents/questions';
import { tabInPath } from '../utils/tab-or-delete-in-path';
import { useDiffAgentActions } from '../context/diff-agent-context';

const options = [
  {
    char: 'a',
    action: 'add path and operation',
  },
  {
    char: 's',
    action: 'skip this example',
  },
];

export function UnmatchedPathQuestion(props: { question: AddPathQuestion }) {
  const [pathPattern, setPathPattern] = useState(
    props.question.diff.closestMatch
  );
  const actions = useDiffAgentActions();
  const [autocompleteKey, setAutocompleteKey] = useState(0);

  const [option, setOption] = useState<string | null>(null);

  const matchesOriginalExample: boolean = useMemo(() => {
    try {
      const { isMatch } = urlPathDiff([pathPattern.trim()]).compareToPath(
        props.question.diff.method,
        props.question.diff.path
      );
      return isMatch;
    } catch (e) {
      return false;
    }
  }, [pathPattern]);

  useInput(
    (input, key) => {
      if (key.tab) {
        setPathPattern(tabInPath(pathPattern, props.question.diff.path));
        setAutocompleteKey((i) => i + 1);
      }
    },
    { isActive: option === 'a' }
  );

  return (
    <Box flexDirection={'column'}>
      <Newline />
      <Divider title="Unmatched Path Observed" />

      <RenderPathMethod
        method={props.question.diff.method}
        path={props.question.diff.path}
      />

      {option === 'a' && (
        <>
          <Box marginRight={1} flexDirection="row">
            <Text bold color="cyan">
              {'  '}
              provide a path pattern:{' '}
            </Text>
          </Box>
          <Box flexDirection="row">
            <Text bold> {' ' + props.question.diff.method.toUpperCase()} </Text>
            <TextInput
              key={autocompleteKey}
              showCursor={true}
              focus={true}
              value={pathPattern}
              onChange={setPathPattern}
              onSubmit={() => {
                const answer: AddPathAnswer = {
                  pathPattern: pathPattern.trim(),
                };
                actions.answer(props.question.uuid, answer);
              }}
            />
          </Box>
          <Box paddingLeft={6} flexDirection="column">
            {matchesOriginalExample && (
              <Box>
                <Text bold color="green">
                  matches example
                </Text>
                <Text bold color="grey">
                  {' '}
                  press{' '}
                  <Text bold color="cyan">
                    (enter)
                  </Text>{' '}
                  to add operation to spec
                </Text>
              </Box>
            )}
            {!matchesOriginalExample && (
              <Box>
                <Text bold color="red">
                  does not match example
                </Text>
                <Text bold color="grey">
                  {' '}
                  update the pattern before adding to spec
                </Text>
              </Box>
            )}
            <Box>
              <Text bold color="grey">
                {'                '}
                press{' '}
                <Text bold color="cyan">
                  (escape)
                </Text>{' '}
                to go back
              </Text>
            </Box>
          </Box>
        </>
      )}

      <AnswerBar
        onSelect={(option) => {
          if (option === 's') {
            return actions.skipInteraction();
          }
          setOption(option);
        }}
        options={options}
        hide={Boolean(option)}
      />
    </Box>
  );
}
