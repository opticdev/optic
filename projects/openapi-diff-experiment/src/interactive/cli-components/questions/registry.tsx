import React from 'react';
import { AnswerQuestionTypes, QuestionsForAgent } from '../../agents/questions';
import { UnmatchedPathQuestion } from './unmatched-path-question';
import { ReviewBodyPropertyPatches } from './review-body-property-patches';

export function componentForQuestion(question: QuestionsForAgent | undefined) {
  if (!question) return null;

  switch (question.type) {
    case AnswerQuestionTypes.AddPath:
      return <UnmatchedPathQuestion key={question.uuid} question={question} />;
    case AnswerQuestionTypes.ReviewPatchesForBodyPropertyDiff:
      return (
        <ReviewBodyPropertyPatches key={question.uuid} question={question} />
      );
    default:
      throw new Error('no registered question / answer component');
  }
}
