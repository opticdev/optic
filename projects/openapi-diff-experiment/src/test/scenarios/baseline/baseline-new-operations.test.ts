import { AnswerQuestionTypes } from '../../../interactive/agents/questions';
import {
  emptySpec,
  oneEndpointExampleSpec,
  oneEndpointExampleSpecWithRequestBody,
} from '../specs';
import { DebugTraffic } from '../traffic';

describe('baseline new operations', () => {
  it('asks question for a new path and can learn response body', async () => {
    const scenario = emptySpec();
    scenario.sendTraffic(
      DebugTraffic('get', '/example').withJsonResponse({ hello: 'world' })
    );

    await scenario.answerQuestion(AnswerQuestionTypes.AddPath, (question) => {
      question.answer = {
        pathPattern: '/example',
      };
    });

    await scenario.waitForEmptyQueue();

    expect(scenario.results.flattenedSpec()).toMatchSnapshot();
  });

  it('asks question for a new path and can learn request body', async () => {
    const scenario = emptySpec();
    scenario.sendTraffic(
      DebugTraffic('get', '/example')
        .withJsonResponse({ hello: 'world' })
        .withJsonRequest({ goodbye: 'earth' })
    );

    await scenario.answerQuestion(AnswerQuestionTypes.AddPath, (question) => {
      question.answer = {
        pathPattern: '/example',
      };
    });

    await scenario.waitForEmptyQueue();

    expect(scenario.results.flattenedSpec()).toMatchSnapshot();
  });

  it('autolearns new http methods on matched paths', async () => {
    const scenario = await oneEndpointExampleSpec();

    scenario.sendTraffic(
      DebugTraffic('put', '/example').withJsonResponse({ hello: 'world' })
    );

    await scenario.waitForEmptyQueue();

    expect(scenario.results.flattenedSpec()).toMatchSnapshot();
  });

  it('autolearns new properties in matched response bodies', async () => {
    const scenario = await oneEndpointExampleSpec();

    scenario.sendTraffic(
      DebugTraffic('get', '/example').withJsonResponse({
        hello: 'world',
        goodbye: 'saturn',
      })
    );

    await scenario.waitForEmptyQueue();

    expect(scenario.results.flattenedSpec()).toMatchSnapshot();
  });

  it('autolearns new properties in matched request bodies', async () => {
    const scenario = await oneEndpointExampleSpecWithRequestBody();

    scenario.sendTraffic(
      DebugTraffic('get', '/example')
        .withJsonRequest({ goodbye: 'earth', colors: ['blue', 'green'] })
        .withJsonResponse({
          hello: 'world',
          goodbye: 'saturn',
        })
    );

    await scenario.waitForEmptyQueue();

    expect(scenario.results.flattenedSpec()).toMatchSnapshot();
  });

  it('autolearns new query parameters on matched endpoints', async () => {
    const scenario = await oneEndpointExampleSpec();

    scenario.sendTraffic(
      DebugTraffic('get', '/example')
        .withJsonResponse({ hello: 'world' })
        .withQuery('limit=50')
    );

    await scenario.waitForEmptyQueue();

    expect(scenario.results.flattenedSpec()).toMatchSnapshot();
  });

  it('autolearns new response status codes on matched paths', async () => {
    const scenario = await oneEndpointExampleSpec();

    scenario.sendTraffic(
      DebugTraffic('get', '/example')
        .withStatusCode('404')
        .withJsonResponse({ error: 'look here...' })
    );

    await scenario.waitForEmptyQueue();

    expect(scenario.results.flattenedSpec()).toMatchSnapshot();
  });
});
