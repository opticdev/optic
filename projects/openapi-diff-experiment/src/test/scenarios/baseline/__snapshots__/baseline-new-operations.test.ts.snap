// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`baseline new operations asks question for a new path and can learn request body 1`] = `
Object {
  "info": Object {
    "title": "Empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "parameters": Array [],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "properties": Object {
                  "goodbye": Object {
                    "type": "string",
                  },
                },
                "required": Array [
                  "goodbye",
                ],
                "type": "object",
              },
            },
          },
          "description": "",
          "required": true,
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "hello": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "hello",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [],
    },
  },
}
`;

exports[`baseline new operations asks question for a new path and can learn response body 1`] = `
Object {
  "info": Object {
    "title": "Empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "hello": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "hello",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [],
    },
  },
}
`;

exports[`baseline new operations autolearns new http methods on matched paths 1`] = `
Object {
  "info": Object {
    "title": "Empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "hello": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "hello",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [],
      "put": Object {
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "hello": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "hello",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
    },
  },
}
`;

exports[`baseline new operations autolearns new properties in matched request bodies 1`] = `
Object {
  "info": Object {
    "title": "Empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "parameters": Array [],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "properties": Object {
                  "colors": Object {
                    "items": Object {
                      "type": "string",
                    },
                    "type": "array",
                  },
                  "goodbye": Object {
                    "type": "string",
                  },
                },
                "required": Array [
                  "goodbye",
                  "colors",
                ],
                "type": "object",
              },
            },
          },
          "description": "",
          "required": true,
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "goodbye": Object {
                      "type": "string",
                    },
                    "hello": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "hello",
                    "goodbye",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [],
      "post": Object {
        "parameters": Array [],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "properties": Object {
                  "goodbye": Object {
                    "type": "string",
                  },
                },
                "required": Array [
                  "goodbye",
                ],
                "type": "object",
              },
            },
          },
          "description": "",
          "required": true,
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "hello": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "hello",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
    },
  },
}
`;

exports[`baseline new operations autolearns new properties in matched response bodies 1`] = `
Object {
  "info": Object {
    "title": "Empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "goodbye": Object {
                      "type": "string",
                    },
                    "hello": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "hello",
                    "goodbye",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [],
    },
  },
}
`;

exports[`baseline new operations autolearns new query parameters on matched endpoints 1`] = `
Object {
  "info": Object {
    "title": "Empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "parameters": Array [
          Object {
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": Object {
              "type": "number",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "hello": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "hello",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [],
    },
  },
}
`;

exports[`baseline new operations autolearns new response status codes on matched paths 1`] = `
Object {
  "info": Object {
    "title": "Empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "hello": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "hello",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
          "404": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "error": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "error",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "parameters": Array [],
    },
  },
}
`;
